### PR TITLE
bcc: add Table.DeleteAll

### PR DESCRIPTION
**Pull request by Pryz**
This PR adds a `DeleteAll` function to be able to easily clear all keys of a table. Idea is to have something like https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#5-clear.

Not quite sure what's the contribution process so let me know if there is something else I have to do here.

### CODE REVIEW

File: bcc/table.go [hunk 1]
```diff
@@ -164,6 +164,24 @@ func (table *Table) Delete(keyStr string) error {
 	return nil
 }
 
+// Delete all keys from the table
+func (table *Table) DeleteAll() error {
+	mod := table.module.p
+	fd := C.bpf_table_fd_id(table.module.p, table.id)
```
Comment (schu) [inline bcc/table.go]: You could also use `mod` here for `table.module.p`

File: bcc/table.go [hunk 2]
```diff
@@ -164,6 +164,24 @@ func (table *Table) Delete(keyStr string) error {
...
+
+	key_size := C.bpf_table_key_size_id(mod, table.id)
+	key := make([]byte, key_size)
+	keyP := unsafe.Pointer(&key[0])
+	res := C.bpf_lookup_elem(fd, keyP, nil)
+	for res = C.bpf_get_next_key(fd, keyP, keyP); res == 0; res = C.bpf_get_next_key(fd, keyP, keyP) {
```
Comment (schu) [inline bcc/table.go]: bcc now has a function [`bpf_get_first_key`](https://github.com/iovisor/bcc/blob/fdf9b08cee7844d9ad86cbb1ab6b5adeff2b0520/src/cc/libbpf.c#L141) which takes care of handling kernel specific peculiarities. I would prefer we wait for the [update to bcc v0.4.0](https://github.com/iovisor/gobpf/pull/73). After that, you could do the `for` as follows:

```
for res = C.bpf_get_first_key(fd, keyP, keySize); res == 0; res = C.bpf_get_next_key(fd, keyP, keyP) {
```

File: bcc/table.go [hunk 3]
```diff
@@ -164,6 +164,24 @@ func (table *Table) Delete(keyStr string) error {
...
+		r, err := C.bpf_delete_elem(fd, keyP)
+		if r != 0 {
+			return fmt.Errorf("Table.DeleteAll: unable to delete element : %v", err)
```
Comment (schu) [inline bcc/table.go]: Stray ` ` before the colon.

**Comment by schu** [comment]
> Not quite sure what's the contribution process so let me know if there is something else I have to do here.

Please prefix your commit message title with `bcc: ...`, otherwise all good :) Thanks!

**Comment by schu** [comment]
gobpf is now updated to bcc v0.4.0.

**Comment by Pryz** [comment]
rebased and updated :)

File: bcc/table.go [hunk 4]
```diff
@@ -164,6 +164,24 @@ func (table *Table) Delete(keyStr string) error {
...
 
+// Delete all keys from the table
+func (table *Table) DeleteAll() error {
+	mod := table.module.p
+	fd := C.bpf_table_fd_id(mod, table.id)
+
+	keySize := C.bpf_table_key_size_id(mod, table.id)
+	key := make([]byte, keySize)
+	keyP := unsafe.Pointer(&key[0])
+	res := C.bpf_lookup_elem(fd, keyP, nil)
```
Comment (schu) [inline bcc/table.go]: Is this `bpf_lookup_elem` still needed?

**Comment by nathanleclaire** [comment]
Why not call it `Clear` for consistency with existing Python bindings?

**Comment by Pryz** [comment]
Any specific reason I guess I could rename it :)

@schu thought ?

**Comment by Pryz** [comment]
@schu I would need some advice for the test. I've pushed a test with a specific bpf program but I'm pretty sure there is an easier way to test that.

**Comment by schu** [comment]
> I would need some advice for the test. I've pushed a test with a specific bpf program but I'm pretty sure there is an easier way to test that.

You could re-use the existing `table1` like follows:

https://github.com/kinvolk/gobpf/commit/544211703374661ef700c4dafba6e2811ea3ba37

I'm fine with both `DeleteAll` and `Clear`, as you prefer.

**Comment by Pryz** [comment]
Oh nice. I've totally missed the `Set` function. Test updated !

**Comment by Pryz** [comment]
Thanks !

**Pull request merged**